gcr.go

```
@@ -0,0 +1,121 @@
+
+// ref: https://stackoverflow.com/questions/34037256/does-google-container-registry-support-docker-remote-api-v2/34046435#34046435
+func fetchBearer(repo string, image string) (string, error) {
+	token, err := exec.Command("gcloud", "auth", "print-access-token").Output() // run gcloud command
+	if err != nil {
+		return "", err
+	}
+
+	url := constructGCRAuthURL(repo, image)
+	body, err := httpGet(url, "", &BasicAuthInfo{
+		Username: "_token",
+		Password: strings.TrimSpace(string(token)), // need to remove trailing new line character
+	})
+
+	var resp gcrAuthResponse
+	if err := json.Unmarshal([]byte(body), &resp); err != nil {
+		return "", err
+	}
+	return resp.Token, nil
+}
+
+func constructGCRAuthURL(repo string, image string) string {
+	u := url.URL{
+		Scheme: "https",
+		Path:   path.Join(repo, "/v2/token"),
+	}
+	q := u.Query()
+	q.Set("scope", fmt.Sprintf("repository:%s:pull", image))
+	u.RawQuery = q.Encode()
+
+	return u.String()
+}
+
+func constructGCRAPIURL(repo string, image string) string {
+	u := url.URL{
+		Scheme: "https",
+		Path:   path.Join(repo, "v2", image, "tags/list"),
+	}
+	return u.String()
+}
+
+func parseGCRTagsResponse(manifests gcrTagsResponse) (gcrImages, error) {
+	gcrImages := gcrImages{}
+	for _, manifest := range manifests.Manifest {
+		var imageDetail gcrImageDetail
+		if err := json.Unmarshal([]byte(manifest), &imageDetail); err != nil {
+			return nil, err
+		}
+		gcrImages = append(gcrImages, imageDetail)
+	}
+	return gcrImages, nil
+}
+
+func extractGCRTagNames(images gcrImages) []string {
+	tags := []string{}
+	sort.Slice(images, func(i, j int) bool {
+		return images[i].TimeCreatedMs > images[j].TimeCreatedMs
+	}) // sort Newset -> Oldest
+
+	for _, image := range images {
+		for _, tag := range image.Tag {
+			tags = append(tags, tag)
+		}
+	}
+	return tags
+}
+
+func retrieveFromGCR(repo string, image string) ([]string, error) {
+	bearer, err := fetchBearer(repo, image)
+	if err != nil {
+		return nil, err
+	}
```

The public repository doesn't require an access token, so it should be continued even if it failed to get the token.

The following command is expected to succeed but now fails.

```console
$ /home/bgpat/bin/dockertags gcr.io/google_containers/hyperkube
exit status 1
```

---

I ignore the error here in 872f215